### PR TITLE
Sb eln export multiple projects

### PR DESCRIPTION
**Pull request by SteffenBrinckmann**
@jmurugan-fzj This is the branch with the new and updated exportELN function in inputOutput.py. The inputELN function is not updated yet, hence not a finalized PR

### CODE REVIEW

**Comment by SteffenBrinckmann** [comment]
@jmurugan-fzj I had to change data_upload_task.py since the argument was wrong. Should I have done that? I guess that made the pytest fail.
Otherwise, this is ready for review

File: pasta_eln/inputOutput.py [hunk 1]
```diff
@@ -346,64 +312,66 @@ def processPart(part:dict[str,str]) -> int:
...
-  def separate(doc: dict[str,Any]) -> tuple[str, dict[str,Any], dict[str,Any]]:
+  def separate(doc: dict[str,Any], dirNameProject_: str, filesNotInProject:list[str]) -> tuple[str, dict[str,Any]]:
     """ separate document into
     - main information (for all elns) and
     - supplemental information (specific to PastaELN)
+
+    Args:
+      doc (dict): document to separate into main and supplemental information
+      dirNameProject_ (str): name of the project directory
+      filesNotInProject (list): list of path-strings of files that are not in project folder hierarchy
```
Comment (jmurugan-fzj) [inline pasta_eln/inputOutput.py:337]: Did you mean an exclude list? The items to be excluded when exporting the project? 
If so, the following, would it not be more appropriate?

excludeItems (list): list of items to be excluded from the exported ELN file.

File: pasta_eln/inputOutput.py [hunk 2]
```diff
@@ -346,64 +312,66 @@ def processPart(part:dict[str,str]) -> int:
...
-
-  def separate(doc: dict[str,Any]) -> tuple[str, dict[str,Any], dict[str,Any]]:
+  def separate(doc: dict[str,Any], dirNameProject_: str, filesNotInProject:list[str]) -> tuple[str, dict[str,Any]]:
     """ separate document into
     - main information (for all elns) and
     - supplemental information (specific to PastaELN)
+
+    Args:
+      doc (dict): document to separate into main and supplemental information
+      dirNameProject_ (str): name of the project directory
```
Comment (jmurugan-fzj) [inline pasta_eln/inputOutput.py:336]: change dirNameProject_ to dirNameProject?

File: pasta_eln/printer.py [hunk 3]
```diff
@@ -15,24 +15,23 @@ def createQRcodeSheet(fileName:str="../qrCodes.pdf") -> None:
   import qrcode
   import numpy as np
   from PIL import Image
-  from commonTools import commonTools as cT  # don't import globally since it disturbs translation
-  img = qrcode.make(cT.uuidv4(),
-                    error_correction=qrcode.constants.ERROR_CORRECT_M)
-  size = img.size[0]
+  #from commonTools import commonTools as cT  # don't import globally since it disturbs translation
```
Comment (jmurugan-fzj) [inline pasta_eln/printer.py:18]: remove the comments if not needed!

File: pasta_eln/serverActions.py [hunk 4]
```diff
@@ -370,9 +370,8 @@ def backupCouchDB(location:str='', userName:str='', password:str='') -> None:
   timestamp = datetime.now().isoformat().split('.')[0].replace('-','').replace(':','')
   zipFileName = 'couchDB_backup_'+location.replace('.','')+'_'+timestamp
   print(f'Create zip-file {zipFileName}.zip')
-  databases = resp.json()
   with ZipFile(f'{zipFileName}.zip', 'w', compression=ZIP_DEFLATED) as zipFile:
-    for database in databases:
+    for database in ['pasta_tutorial2']: #databases:
```
Comment (jmurugan-fzj) [inline pasta_eln/serverActions.py:374]: Is this intended? is 'pasta_tutorial2' a standard database?

**Review by jmurugan-fzj — CHANGES_REQUESTED** [review]

**Comment by jmurugan-fzj** [comment]
> exportELN function

@SteffenBrinckmann Could you quickly describe the signature of the exportELN function with the new changes as part of the PR? I will create a separate PR for the data_upload_task.py changes then!

File: pasta_eln/gui.py [hunk 5]
```diff
@@ -141,9 +142,16 @@ def execute(self, command: list[Any]) -> None:
...
+      fileName = QFileDialog.getSaveFileName(self, 'Save project into .eln file', str(Path.home()), '*.eln')[0]
       if fileName != '':
-        status = exportELN(self.comm.backend, self.comm.projectID, fileName)
+        status = exportELN(self.comm.backend, [self.comm.projectID], fileName)
+        showMessage(self, 'Finished', status, 'Information')
+    elif command[0] is Command.EXPORT_ALL:
+      fileName = QFileDialog.getSaveFileName(self, 'Save everything to .eln file', str(Path.home()), '*.eln')[0]
+      if fileName != '':
+        allProjects = [i['id'] for i in self.comm.backend.db.getView('viewDocType/x0')]
+        print(allProjects)
```
[the remark this answers is not in the document]
Comment (SteffenBrinckmann) [reply pasta_eln/gui.py:153]: No, I removed it

File: pasta_eln/inputOutput.py (hunk 2 above)
Comment (SteffenBrinckmann) [reply pasta_eln/inputOutput.py:336]: Did that

File: pasta_eln/serverActions.py (hunk 4 above)
Comment (SteffenBrinckmann) [reply pasta_eln/serverActions.py:374]: No, sorry.

File: pasta_eln/inputOutput.py (hunk 1 above)
Comment (SteffenBrinckmann) [reply pasta_eln/inputOutput.py:337]: I changed the code to make it more clear

File: pasta_eln/printer.py (hunk 3 above)
Comment (SteffenBrinckmann) [reply pasta_eln/printer.py:18]: I commented the entire code, such that it can be used in the future when the issue is fixed.

**Comment by SteffenBrinckmann** [comment]
@jmurugan-fzj This should be the signature of the export-function
def exportELN(backend:Backend, projectIDs:list[str], fileName:str, dTypes:list[str]=[], verbose:bool=False) -> str:
  """
  export eln to file

  Args:
    backend (backend): PASTA backend instance
    projectIDs (list): list of docIds of projects
    fileName (str): fileName which to use for saving
    dTypes (list): list of strings which should be included in the output, alongside folders x0 & x1; empty list=everything is exported
    verbose (bool): verbose

  Returns:
    str: report of exportation
  """

**Review by jmurugan-fzj — APPROVED** [review]

**Pull request merged**